File: android/feature/onboarding/src/main/kotlin/com/fatwabot/feature/onboarding/OnboardingViewModel.kt
package com.fatwabot.feature.onboarding

import androidx.lifecycle.ViewModel
import com.fatwabot.core.common.OnboardingCompletionStore
import com.fatwabot.core.network.AccountProvider
import com.fatwabot.core.network.AccountServicing
import com.fatwabot.core.network.ProviderCredentialProviding
import dagger.hilt.android.lifecycle.HiltViewModel
import javax.inject.Inject
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow

/**
 * The screens of the value-first onboarding flow (docs/features/onboarding.md).
 * SIGN_IN is last on purpose — accounts are optional and must never gate the
 * app (docs/features/accounts.md).
 */
enum class OnboardingStep {
    WELCOME, HIGHLIGHTS, LOCATION_PRIMING, NOTIFICATION_PRIMING, SIGN_IN
}

/**
 * State machine for onboarding. Unlike iOS, Android's runtime-permission
 * dialogs can only be launched from a Composable (`rememberLauncherForActivityResult`),
 * so this ViewModel owns navigation + completion-persistence only — the
 * Composable itself triggers the actual OS permission prompts and then calls
 * back into [advance]/[finish].
 */
@HiltViewModel
class OnboardingViewModel @Inject constructor(
    private val completionStore: OnboardingCompletionStore,
    private val account: AccountServicing,
    private val credentials: ProviderCredentialProviding,
) : ViewModel() {
    private val _step = MutableStateFlow(OnboardingStep.WELCOME)
    val step: StateFlow<OnboardingStep> = _step.asStateFlow()

    private val _isSigningIn = MutableStateFlow(false)
    val isSigningIn: StateFlow<Boolean> = _isSigningIn.asStateFlow()

    private val _signInFailed = MutableStateFlow(false)
    val signInFailed: StateFlow<Boolean> = _signInFailed.asStateFlow()

    private var finished = false

    /** Providers whose SDK is actually wired on this build (Apple has none on Android). */
    val signInOptions: List<AccountProvider> =
        listOf(AccountProvider.APPLE, AccountProvider.GOOGLE).filter(credentials::isConfigured)

    /** Sequences forward; finishes onboarding once past the last step. */
    fun advance() {
        val next = OnboardingStep.entries.getOrNull(_step.value.ordinal + 1)
        when {
            next == null -> finish()
            // Nothing to offer — don't show an empty screen.
            next == OnboardingStep.SIGN_IN && signInOptions.isEmpty() -> finish()
            else -> _step.value = next
        }
    }

    /** "Not now" / "Skip" — never triggers a permission request, just sequences forward. */
    fun skip() = advance()

    /**
     * Links the anonymous identity to [provider], so everything accumulated
     * during onboarding carries over. Returns true when onboarding is complete;
     * a cancel/failure keeps the user on the step instead of silently dropping
     * them into the app.
     */
    suspend fun signIn(provider: AccountProvider): Boolean {
        _isSigningIn.value = true
        _signInFailed.value = false
        val succeeded = runCatching {
            val token = credentials.identityToken(provider)
            account.link(provider, token)
        }.isSuccess
        _isSigningIn.value = false
        if (succeeded) {
            finish()
        } else {
            _signInFailed.value = true
        }
        return succeeded
    }

    /** "Continue as guest" — the app is fully usable without an account. */
    fun continueAsGuest() = finish()

    fun finish() {
        if (finished) return
        finished = true
        completionStore.markCompleted(System.currentTimeMillis() / 1000)
    }
}

File: ios/Packages/AppFeatures/Sources/OnboardingFeature/OnboardingViewModel.swift
import CoreKit
import Observation

/// State machine for the 4-screen value-first onboarding flow
/// (docs/features/onboarding.md). Permission requests are injected as plain
/// closures rather than depending on PrayerFeature's LocationProviding /
/// PrayerNotificationScheduling directly — ADR-0010 forbids feature -> feature
/// dependencies, so the App composition root wires these to the real
/// M1 permission-request code paths; this package never touches CoreLocation
/// or UNUserNotificationCenter itself.
public enum OnboardingStep: Int, CaseIterable, Sendable {
    case welcome
    case highlights
    case locationPriming
    case notificationPriming
    /// Last on purpose — accounts are optional and must never gate the app
    /// (docs/features/accounts.md); the user has already seen the value.
    case signIn
}

/// A sign-in provider offered during onboarding. Kept as a plain value type so
/// OnboardingFeature doesn't depend on NetworkingKit (ADR-0010) — the App
/// composition root maps `id` back to an `AccountProvider`.
public struct OnboardingSignInOption: Identifiable, Sendable, Equatable {
    public let id: String
    public let titleKey: String
    public let systemImage: String

    public init(id: String, titleKey: String, systemImage: String) {
        self.id = id
        self.titleKey = titleKey
        self.systemImage = systemImage
    }
}

@MainActor
@Observable
public final class OnboardingViewModel {
    public private(set) var step: OnboardingStep = .welcome
    /// Providers whose SDK is actually wired on this platform/build.
    public let signInOptions: [OnboardingSignInOption]
    public private(set) var isSigningIn = false
    public private(set) var signInFailed = false

    private let requestLocation: @Sendable () async -> Void
    private let requestNotifications: @Sendable () async -> Void
    private let performSignIn: @Sendable (String) async -> Bool
    private let completionStore: OnboardingCompletionStore
    private let onFinished: @Sendable () -> Void

    public init(
        requestLocation: @escaping @Sendable () async -> Void,
        requestNotifications: @escaping @Sendable () async -> Void,
        signInOptions: [OnboardingSignInOption] = [],
        performSignIn: @escaping @Sendable (String) async -> Bool = { _ in false },
        completionStore: OnboardingCompletionStore,
        onFinished: @escaping @Sendable () -> Void
    ) {
        self.requestLocation = requestLocation
        self.requestNotifications = requestNotifications
        self.signInOptions = signInOptions
        self.performSignIn = performSignIn
        self.completionStore = completionStore
        self.onFinished = onFinished
    }

    public func advance() {
        guard let next = OnboardingStep(rawValue: step.rawValue + 1) else {
            finish()
            return
        }
        // Nothing to offer (e.g. no provider SDK wired) — don't show an empty
        // screen, just finish.
        if next == .signIn, signInOptions.isEmpty {
            finish()
            return
        }
        step = next
    }

    /// "Allow" on the location-priming screen: triggers the real OS prompt,
    /// then advances regardless of the outcome — denial still reaches Home
    /// via the existing manual-city fallback (docs/features/prayer.md).
    public func allowLocation() async {
        await requestLocation()
        advance()
    }

    /// "Allow" on the notification-priming screen: triggers the real OS
    /// prompt, then moves on regardless of outcome.
    public func allowNotifications() async {
        await requestNotifications()
        advance()
    }

    /// Taps a provider on the sign-in screen. Success finishes onboarding;
    /// failure keeps the user here with a message (cancelling is silent — the
    /// closure reports false and we simply stay put). Either way the account
    /// is optional: "Continue as guest" is always available.
    public func signIn(with optionId: String) async {
        isSigningIn = true
        signInFailed = false
        let succeeded = await performSignIn(optionId)
        isSigningIn = false
        if succeeded {
            finish()
        } else {
            signInFailed = true
        }
    }

    /// "Continue as guest" — the app is fully usable without an account.
    public func continueAsGuest() {
        finish()
    }

    /// "Not now" / "Skip" on any screen — never calls the permission API,
    /// just sequences forward (or finishes, from the last screen).
    public func skip() {
        advance()
    }

    private func finish() {
        completionStore.markCompleted()
        onFinished()
    }
}
